ext/datadog_cov/datadog_cov.c

```diff
@@ -18,30 +27,75 @@ char *ruby_strndup(const char *str, size_t size)
   return dup;
 }
 
+static VALUE just_return_nil(VALUE _not_used_self, VALUE _not_used_exception)
+{
+  return Qnil;
+}
+
+// Equivalent to Ruby "begin/rescue nil" call, where we call a C function and
+// swallow the exception if it occurs - const_source_location often fails with
+// exceptions for classes that are defined in C or for anonymous classes.
+static VALUE rescue_nil(VALUE (*function_to_call_safely)(VALUE), VALUE function_to_call_safely_arg)
+{
+  return rb_rescue2(
+      function_to_call_safely,
+      function_to_call_safely_arg,
+      just_return_nil,
+      Qnil,
+      rb_eException, // rb_eException is the base class of all Ruby exceptions
+      0              // Required by API to be the last argument
+  );
+}
+
 // Data structure
 struct dd_cov_data
 {
+  // Ruby hash with filenames impacted by the test.
+  VALUE impacted_files;
+
+  // Root is the path to the root folder of the project under test.
+  // Files located outside of the root are ignored.
   char *root;
   long root_len;
 
+  // Ignored path contains path to the folder where bundled gems are located if
+  // gems are installed in the project folder.
   char *ignored_path;
   long ignored_path_len;
 
-  VALUE coverage;
-
+  // Line tracepoint optimisation: cache last seen filename pointer to avoid
+  // unnecessary string comparison if we stay in the same file.
   uintptr_t last_filename_ptr;
 
+  // Line tracepoint can work in two modes: single threaded and multi threaded
+  //
+  // In single threaded mode line tracepoint will only cover the thread that started the coverage.
+  // This mode is useful for testing frameworks that run tests in multiple threads.
+  // Do not use single threaded mode for Rails applications unless you know that you
+  // don't run any background threads.
+  //
+  // In multi threaded mode line tracepoint will cover all threads. This mode is enabled by default
+  // and is recommended for most applications.
+  int threading_mode;
```

Comment:
Minor: I don't use them very often myself (yet...?) but this may be a good candidate for using an `enum`.